src/toxcore/friend_connection/mod.rs

```diff
@@ -1,4 +1,1070 @@
+            friend.ping_received_time = Some(clock_now());
+        }
+    }
+
+    /// Handle the stream of found DHT `PublicKey`s.
+    fn handle_dht_pk(&self, dht_pk_rx: mpsc::UnboundedReceiver<(PublicKey, PublicKey)>) -> impl Future<Item = (), Error = RunError> + Send {
+        let dht = self.dht.clone();
+        let net_crypto = self.net_crypto.clone();
+        let onion_client = self.onion_client.clone();
+        let friends = self.friends.clone();
+        let tcp_connections = self.tcp_connections.clone();
+        dht_pk_rx
+            .map_err(|()| unreachable!("rx can't fail"))
+            .for_each(move |(real_pk, dht_pk)| {
+                if let Some(friend) = friends.write().get_mut(&real_pk) {
+                    friend.dht_pk_time = Some(clock_now());
+
+                    if friend.dht_pk != Some(dht_pk) {
+                        info!("Found a friend's DHT key");
+
+                        let kill_future = if let Some(dht_pk) = friend.dht_pk {
+                            dht.remove_friend(dht_pk);
+
+                            let kill_connection_future = net_crypto.kill_connection(real_pk)
+                                .then(|res| match res {
+                                    Err(ref e) if *e.kind() == KillConnectionErrorKind::NoConnection => Ok(()),
+                                    res => res,
+                                })
+                                .map_err(|e| e.context(RunErrorKind::KillConnection).into());
+                            // TODO: handle error properly after migrating the TCP client to failure
+                            let tcp_remove_future = tcp_connections.remove_connection(dht_pk).then(|_| Ok(()));
+
+                            Either::A(kill_connection_future.join(tcp_remove_future).map(|_| ()))
+                        } else {
+                            Either::B(future::ok(()))
+                        };
+
+                        friend.dht_pk = Some(dht_pk);
+
+                        dht.add_friend(dht_pk);
+                        net_crypto.add_connection(real_pk, dht_pk);
+                        onion_client.set_friend_dht_pk(real_pk, dht_pk);
+
+                        kill_future
+                    } else {
+                        Either::B(future::ok(()))
+                    }
+                } else {
+                    Either::B(future::ok(()))
+                }
+            })
+    }
+
+    /// Handle the stream of found IP addresses.
+    fn handle_friend_saddr(&self, friend_saddr_rx: mpsc::UnboundedReceiver<PackedNode>) -> impl Future<Item = (), Error = RunError> + Send {
+        let net_crypto = self.net_crypto.clone();
+        let friends = self.friends.clone();
+        friend_saddr_rx
+            .map_err(|()| unreachable!("rx can't fail"))
+            .for_each(move |node| {
+                if let Some(friend) = friends.write().values_mut().find(|friend| friend.dht_pk == Some(node.pk)) {
```

Comment:
If `friend: Copy` you will make a shallow copy in `if let Some(friend)`

Reply:
It doesn't implement Copy, and even if it were it'd still be a reference.